File: python/src/commands/set.py
from .base import BaseCommand
from src.resp.serializer import RespSerializer
from datetime import datetime
from time import time

class Set(BaseCommand):
  @classmethod
  def command(cls) -> bytes:
    return b'set'
  
  def run(self) -> bytes:
    key: bytes = self.args[0]
    value: bytes = self.args[1]
    if len(self.args) > 2:
      expire_type: bytes = self.args[2]
      if expire_type:
        expire_type = expire_type.lower()
        expire_time: bytes = self.args[3]
        if not expire_time:
          return RespSerializer.serialize(Exception('No Expire time'))

      self._register_expire_keys(key=key, expire_type=expire_type, expire_time=expire_time)

    self.store[key] = value
    return RespSerializer.serialize(b'OK')
  
  def _register_expire_keys(self, key: bytes, expire_type: bytes, expire_time: bytes):
    ttl: int = int(expire_time.decode('UTF-8'))
    expire_at: int = 0
    if expire_type == b'ex':
      expire_at = (time() + ttl) * 1000
    elif expire_type == b'px':
      expire_at = time() * 1000 + ttl
    elif expire_type == b'exat':
      expire_at = ttl * 1000
    elif expire_type == b'pxat':
      expire_at = ttl
    
    self.expire_store[key] = expire_at

File: ruby/lib/redis/commands/set.rb
# frozen_string_literal: true
#
require_relative './base'

module Redis
  module Commands
    class Set < Base
      EX = 'EX'
      PX = 'PX'
      EXAT = 'EXAT'
      PXAT = 'PXAT'

      def self.command
        'set'
      end
      
      def run
        @mutex.synchronize do
          key = args[0]
          value = args[1]
          # Add the key to the expire keys hash
          if args.size > 2
            expire_type = args[2]
            expire_value = args[3].to_i
            add_expire_keys(key, expire_type, expire_value.to_i)
          end
          
          @store[key] = value
          Resp::Serializer.serialize(:OK)
        end
      end

      private

      def add_expire_keys(expire_key, expire_type, expire_value)
        if expire_value.negative?
          return '-ERR value is not an integer or out of range'
        end
        
        standard_expired_value = \
          case expire_type.upcase
          when EX # seconds
            (Time.now.to_i + expire_value) * 1000
          when PX # milliseconds
            Time.now.to_i * 1000 + expire_value
          when EXAT # Unix time in seconds
            expire_value * 1000
          when PXAT # Unix time in milliseconds
            expire_value
          else
            return "-ERR Unknown expire type options #{expire_type}"
          end
        
        @expire_keys[expire_key] = standard_expired_value
      end
    end
  end
end